lib/presto-client/index.js

```diff
@@ -220,167 +229,212 @@ Client.prototype.statementResource = function(opts) {
-                 "nextUri": "http://localhost:8080/v1/statement/20140120_032523_00000_32v8g/2",
-                 "partialCancelUri": "http://10.0.0.0:8080/v1/stage/20140120_032523_00000_32v8g.0",
-                 "infoUri": "http://localhost:8080/v1/query/20140120_032523_00000_32v8g",
-                 "id": "20140120_032523_00000_32v8g"
-             }
-            */
-            /*
-             * 2nd time
-             {
-                 "stats": {
-                     // ....
-                     "state": "FINISHED",
-                 },
-                 "columns": [ { "type": "bigint", "name": "cnt" } ],
-                 "infoUri": "http://localhost:8080/v1/query/20140120_032523_00000_32v8g",
-                 "id": "20140120_032523_00000_32v8g"
-             }
-            */
-            if (cancel_checker && cancel_checker()) {
-                client.request({ method: 'DELETE', path: next_uri }, function(error, code, data){
-                    if (error || code !== 204) {
-                        error_callback({message: "query fetch canceled, but Presto query cancel may fail", error: error, code: code});
-                    } else {
-                        error_callback({message: "query fetch canceled by operation"});
-                    }
-                });
+        current_req = client.request(next_uri, function(error, code, response){
+            // same as above, we have already timed out and shown an error, so abort out
+            if (timed_out) {
+                return;
+            } else if ([502, 503, 504].includes(code)) {
+                setTimeout(function(){
+                    fetch_next(next_uri);
+                }, Math.floor(Math.random() * 51) + 50);
+                if (retry_callback) {
+                    retry_callback();
+                }
                 return;
             }
-            client.request(next_uri, function(error, code, response){
-                if (error || response.error) {
+
+            if (error || (response && response.error)) {
+                if (first_request) {
+                    var message = "execution error" + (response && response.length > 0 ? ":" + response : "");
+                    if (response && response.error && response.error.message)
+                        message = response.error.message;
+                    error_callback({message, error: (error || response.error), code});
+                } else {
                     error_callback(error || response.error);
-                    return;
                 }
+                clear_timeout();
+                return;
+            }
 
-                if (state_callback && (last_state !== response.stats.state || enable_verbose_state_callback)) {
-                    state_callback(null, response.id, response.stats);
-                    last_state = response.stats.state;
-                }
+            if (first_request && (!response.id || !response.nextUri || !response.infoUri)) {
```

Comment:
Could you explain why this is checked only when it's the first request?

Reply:
Similar to above, this block is replicating this from master:

https://github.com/tagomoris/presto-client-node/blob/84f76d981482c5dd710a147ebfe89efa1731d85f/lib/presto-client/index.js#L260-L270

where this only gets run on that first POST request, and then all subsequent GET requests do not have it. If you wanted me to apply this to all requests, then could only reasonably check for `response.id` for existing, as I think per the REST API docs, the `nextUri` could be undefined on the final GET payload, while `infoUri` is not mentioned so I'm not sure about that.